packages/driver/src/cy/commands/navigation.ts

```diff
@@ -129,25 +117,39 @@ const navigationChanged = (Cypress, cy, state, source, arg) => {
   }
 
   // start storing the history entries
-  const urls = state('urls') || []
+  let urls = state('urls') || []
+  let urlPosition = state('urlPosition')
+
+  if (urlPosition === undefined) {
+    urlPosition = -1
+  }
 
-  const previousUrl = _.last(urls)
+  const previousUrl = urls[urlPosition]
 
-  // ensure our new url doesnt match whatever
+  // ensure our new url doesn't match whatever
   // the previous was. this prevents logging
-  // additionally when the url didnt actually change
+  // additionally when the url didn't actually change
   if (url === previousUrl) {
     return
   }
 
   // else notify the world and log this event
   Cypress.action('cy:url:changed', url)
 
-  urls.push(url)
+  const navHistoryDelta = state('navHistoryDelta')
 
-  state('urls', urls)
+  if (navHistoryDelta) {
+    urlPosition = urlPosition + navHistoryDelta
+    state('navHistoryDelta', undefined)
+  } else {
+    urls = urls.slice(0, urlPosition + 1)
+    urls.push(url)
+    urlPosition = urlPosition + 1
+  }
```

Comment:
This could benefit from a comment about what it's doing.

Reply:
Good call: [`6ba0e0c` (#18996)](https://github.com/cypress-io/cypress/pull/18996/commits/6ba0e0cc601e43a7c12768804f094828d9b971a3)